src/v/storage/disk_log_impl.h

```diff
@@ -243,11 +246,36 @@ class disk_log_impl final : public log {
     // Returns if the update actually took place.
     ss::future<bool> update_start_offset(model::offset o);
 
-    ss::future<compaction_result> compact_adjacent_segments(
+    // Finds a range of adjacent segments that can be compacted together.
+    // A valid segment range consists of segments with the same raft term, and a
+    // combined size less than max_compacted_segment_size.
+    //
+    // Returns std::nullopt if a valid range of two segments could not be found.
+    // Otherwise, a pair of iterators to the segments.
+    std::optional<std::pair<segment_set::iterator, segment_set::iterator>>
+    find_adjacent_compaction_range(const compaction_config& cfg);
+
+    // Requests compaction of adjacent segments per the max_collectible_offset
+    // in the compaction_config.
+    //
+    // Returns std::nullopt if an adjacent pair could not be found for adjacent
+    // compaction, or a compaction_result. This result should also have its
+    // did_compact() function checked to verify whether adjacent segment
+    // compaction actually occurred.
+    ss::future<std::optional<compaction_result>>
+    compact_adjacent_segments(storage::compaction_config cfg);
+
+    // Performs compaction of adjacent segments. This pair of segments is
```

Comment:
This is so confusing because the naming suggests it compacts a range but then the new comment says it compacts a pair ... but then the pair given as input is an begin/end iterator pair...

Tried to read the code, indeed we limit the find segment range to 2 for no documented reason (couldn't find it).

The compact_adjacent_segments also restricts it's input to be 2 segments but it does so in an assert at the end of the function. Why asserts on the inputs aren't at the top?

```
vassert(segments.size() == 2, "Cannot compact more than two segments");
```

I'd probably change the wording to talk about the range and just add at the end a note that currently there is a constraint on the range size to be equal to 2.

Reply:
>This is so confusing because the naming suggests it compacts a range but then the new comment says it compacts a pair ... but then the pair given as input is an begin/end iterator pair...

I agree, it feels like this code was changed to start to move adjacent compaction away from just two adjacent segments into a range of segments, but as it is today this `std::pair<>` of iterators still refers to just two segments, and not a `begin/end` range.

> Why asserts on the inputs aren't at the top?

Also a good question. I can move the assert if we think that improves readability.